core/WASM_core.md

```diff
@@ -0,0 +1,352 @@
+# Summary
+
+This proposal intends to outline the shape of the interfaces and implementation work necessary
+to make Core WASM-compilable.
+
+# Motivation
+
+A few problems a WASM Core Can solve, in no particular order:
+
+* Allow Go and Java to be ported to run on top of Core. *Potentially* without the downsides of CGo
+  and JNI. More on that later.
+* No longer any need to ship compiled Core binaries of various architectures with our existing SDKs.
+  This is both a nice plus for us, in terms of reducing CI complexity, but also a substantial size
+  reduction in our packages.
+* Workers in the browser
+* Hosted workers (also possible without, but with, there are some nice benefits)
+* A simpler way to manage callbacks into the lang layer, enabling things like user-defined metrics
+  exporters, lower-level interceptors, etc. This is certainly doable without WASM, but is probably
+  a bit easier (and, necessary) with it.
+* Auto-updates to core logic. This isn't detailed in this doc, but is a potentially very compelling
+  feature. Imagine a worker that can update core logic without needing to be restarted.
+* Allow the use of lang-native gRPC clients (though core can still provide things like retry and 
+  backoff logic, etc.)
+
+
+# Detailed design
+
+The first and largest chunk of work involved in implementation is more or less pure grunt work. 
+I need to switch Core to compiling with the wasm32-unknown-unknown target, and then fix all the
+compiler errors.
+
+There are a few predictable problems we'll definitely have to deal with:
+
+The largest hiccup is the use of network (gRPC) calls. There is more than one way we could skin this
+cat. Luckily (intentionally), all of this is contained to the `temporal-client` crate.
+
+The next major source of side effects come from the tracing and metrics exporting facilities for
+OTel and Prometheus.
+
+Additionally, there is the code dealing with downloading and starting the CLI / Java test server.
+
+Finally, there may be some issues with other things that in native world use syscalls like atomics,
+locks, random number gen, etc. Luckily my experience so far is that those pretty much all just work
+without any, or fairly minor, changes.
+
+Ultimately we have two strategies available to us. We can create interfaces for Core to call back
+into lang to perform all these functions, or we can attempt to leverage WASI for this functionality.
+WASI is quite immature, but *in theory* it can support these use cases.
```

Comment:
I vote no WASI. A set of well-defined C-style externs for each direction is going to give us the smallest, nicest integration I think.